agent/run_job.go

```diff
@@ -0,0 +1,236 @@
+		if err != nil {
+			r.logger.Error("Metric submission failed to parse %s", r.conf.Job.RunnableAt)
+		} else {
+			r.conf.MetricsScope.Timing("queue.duration", r.startedAt.Sub(runnableAt))
+		}
+	}
+
+	// Start the header time streamer
+	go r.headerTimesStreamer.Run(ctx)
+
+	// Start the log streamer. Launches multiple goroutines.
+	if err := r.logStreamer.Start(ctx); err != nil {
+		return err
+	}
+
+	// Default exit status is no exit status
+	exit := &processExit{}
+
+	// Used to wait on various routines that we spin up
+	var wg sync.WaitGroup
+
+	// Set up a child context for helper goroutines related to running the job.
+	cctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	defer r.cleanup(cctx, &wg, exit)
+
+	// Before executing the bootstrap process with the received Job env, execute the pre-bootstrap hook (if present) for
+	// it to tell us whether it is happy to proceed.
+	if hook, _ := hook.Find(r.conf.AgentConfiguration.HooksPath, "pre-bootstrap"); hook != "" {
+		// Once we have a hook any failure to run it MUST be fatal to the job to guarantee a true positive result from the hook
+		ok, err := r.executePreBootstrapHook(ctx, hook)
+		if !ok {
+			// Ensure the Job UI knows why this job resulted in failure
+			r.logStreamer.Process([]byte("pre-bootstrap hook rejected this job, see the buildkite-agent logs for more details"))
+			// But disclose more information in the agent logs
+			r.logger.Error("pre-bootstrap hook rejected this job: %s", err)
+
+			exit.Status = "-1"
+			exit.SignalReason = "agent_refused"
+
+			return nil
+		}
+	}
+
+	// Kick off log streaming and job status checking when the process starts.
+	wg.Add(2)
+	go r.jobLogStreamer(cctx, &wg)
+	go r.jobCancellationChecker(cctx, &wg)
+
+	exit = r.runJob(cctx) // Ignore gostaticcheck here, the value of exit is captured by the deferred cleanup function above
+
+	return nil
+}
+
+type processExit struct {
+	Status       string
+	Signal       string
+	SignalReason string
+}
+
+func (r *JobRunner) runJob(ctx context.Context) *processExit {
+	exit := &processExit{}
+	// Run the process. This will block until it finishes.
+	if err := r.process.Run(ctx); err != nil {
+		// Send the error as output
+		r.logStreamer.Process([]byte(err.Error()))
+
+		// The process did not run at all, so make sure it fails
+		return &processExit{
+			Status:       "-1",
+			SignalReason: "process_run_error",
+		}
+	}
+	// Intended to capture situations where the job-exec (aka bootstrap) container did not
+	// start. Normally such errors are hidden in the Kubernetes events. Let's feed them up
+	// to the user as they may be the caused by errors in the pipeline definition.
+	if r.cancelled && !r.stopped {
+		k8sProcess, ok := r.process.(*kubernetes.Runner)
+		if ok && k8sProcess.ClientStateUnknown() {
+			r.logStreamer.Process([]byte(
+				"Some containers had unknown exit statuses. Perhaps they were in ImagePullBackOff.",
+			))
+		}
+	}
```

Comment:
I know I wrote this code, but now I wonder if we might want to simply it further to be:
```go
	k8sProcess, ok := r.process.(*kubernetes.Runner)
	if ok && r.cancelled && !r.stopped && k8sProcess.ClientStateUnknown() {
		r.logStreamer.Process([]byte(
			"Some containers had unknown exit statuses. Perhaps they were in ImagePullBackOff.",
		))
	}
```

Reply:
I guess I thought doing the cast was unnecessary if the job was not in the right state. 🤷‍♂️